server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java

```diff
@@ -102,34 +103,46 @@ public void setMappingLimit(int mappingLimit) throws IllegalArgumentException {
 
     @Override
     public Mappings getResolvedMappings(String user, Domain domain) throws ErrorMappingException, RecipientRewriteTableException {
-        return getMappings(Username.fromLocalPartWithDomain(user, domain), mappingLimit);
+        return getMappings(Username.fromLocalPartWithDomain(user, domain), mappingLimit, Optional.empty());
     }
 
-    private Mappings getMappings(Username username, int mappingLimit) throws ErrorMappingException, RecipientRewriteTableException {
+
+    @Override
+    public Mappings getResolvedMappings(String user, Domain domain, EnumSet<Type> mappingTypes) throws ErrorMappingException, RecipientRewriteTableException {
+
+        return getMappings(Username.fromLocalPartWithDomain(user, domain), mappingLimit, Optional.of(mappingTypes));
+    }
+
+    private Mappings getMappings(Username username, int mappingLimit, Optional<EnumSet<Type>> mappingTypes) throws ErrorMappingException, RecipientRewriteTableException {
 
         // We have to much mappings throw ErrorMappingException to avoid
         // infinity loop
         if (mappingLimit == 0) {
             throw new TooManyMappingException("554 Too many mappings to process");
         }
 
-        Mappings targetMappings = mapAddress(username.getLocalPart(), username.getDomainPart().get());
-
+        Domain domain = username.getDomainPart().get();
+        String localPart = username.getLocalPart();
+        Stream<Mapping> targetMappings = mappingTypes.map(Throwing.<EnumSet<Type>, Stream<Mapping>>function(consideredTypes ->
+            mapAddress(localPart, domain).asStream()
+                .filter(mapping -> consideredTypes.contains(mapping.getType())))
+            .sneakyThrow())
+            .orElse(mapAddress(localPart, domain).asStream());
 
         try {
             return MappingsImpl.fromMappings(
-                targetMappings.asStream()
-                    .flatMap(Throwing.function((Mapping target) -> convertAndRecurseMapping(username, target, mappingLimit)).sneakyThrow()));
+                targetMappings
+                    .flatMap(Throwing.function((Mapping target) -> convertAndRecurseMapping(mappingTypes, username, target, mappingLimit)).sneakyThrow()));
```

Comment:
we could remove checked exception from the private method to make this call shorter

Reply:
I am not sure to understand what you want to do exactly, and how to do it without breaking the behaviour of the code relying on those exceptions down the call chain.
This PR only added a parameter to this method. 
So it can be a refactoring but i am not sure to see the value of it.

Reply:
You are right about one thing: we can't throw a checked exceptions without changing its type without using "sneaky throw". However, my comment still stands: it's hard to read.
There are several solutions to that: extract the function into a variable, using sneaky-throw in the sub-method, etc.